systemtest/benchtest/expvarmetrics/metrics.go

```diff
@@ -0,0 +1,253 @@
+	c.processMetric(Bytes, e.UncompressedBytes)
+	c.processMetric(TotalEvents, e.TotalEvents)
+	c.processMetric(ActiveEvents, e.ActiveEvents)
+	c.processMetric(RSSMemoryBytes, e.RSSMemoryBytes)
+	c.processMetric(AvailableBulkRequests, e.AvailableBulkRequests)
+	c.processMetric(TransactionsProcessed, e.TransactionsProcessed)
+	c.processMetric(SpansProcessed, e.SpansProcessed)
+	c.processMetric(MetricsProcessed, e.MetricsProcessed)
+	c.processMetric(ErrorsProcessed, e.ErrorsProcessed)
+	c.processMetric(ErrorElasticResponses, e.ErrorElasticResponses)
+	c.processMetric(ErrorOTLPTracesResponses, e.ErrorOTLPTracesResponses)
+	c.processMetric(ErrorOTLPMetricsResponses, e.ErrorOTLPMetricsResponses)
+	c.processMetric(NumGC, int64(e.NumGC))
+	c.processMetric(PauseTotalNs, int64(e.PauseTotalNs))
+	c.processMetric(MemAllocs, int64(e.Mallocs))
+	c.processMetric(MemBytes, int64(e.TotalAlloc))
+	c.processMetric(HeapAlloc, int64(e.HeapAlloc))
+	c.processMetric(HeapObjects, int64(e.HeapObjects))
+}
+
+func (c *Collector) processMetric(m Metric, val int64) {
+	if _, ok := c.metrics[m]; !ok {
+		c.metrics[m] = &AggregateStats{}
+	}
+	c.updateMetric(c.metrics[m], val)
+
+	if watcher, ok := c.watchers[m]; ok && watcher.validator(val) {
+		watcher.notifyChan <- true
+		delete(c.watchers, m)
+	}
+}
+
+func (c *Collector) updateMetric(stats *AggregateStats, value int64) {
+	// Initialize for first value
+	if stats.sampleCount == 0 {
+		stats.First = value
+		stats.Min = value
+		stats.Max = value
+	}
+
+	stats.Last = value
+	stats.sampleCount += 1
+	stats.Mean += (float64(value) - stats.Mean) / float64(stats.sampleCount)
+	if stats.Min > value {
+		stats.Min = value
+	}
+	if stats.Max < value {
+		stats.Max = value
+	}
+}
+
+func (c *Collector) rejectWatchers() {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	for _, watcher := range c.watchers {
+		watcher.notifyChan <- false
+	}
+}
+
+func (c *Collector) start(ctx context.Context) error {
+	var first expvar
+	err := queryExpvar(ctx, &first, c.serverURL)
+	if err != nil {
+		return err
+	}
+	c.addExpvar(first)
+
+	outChan, errChan := run(ctx, c.serverURL, c.period)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				c.rejectWatchers()
+				return
+			case <-errChan:
+				c.rejectWatchers()
+				return
+			case m := <-outChan:
+				c.addExpvar(m)
+			}
+		}
+	}()
+	return nil
+}
+
+func run(ctx context.Context, serverURL string, period time.Duration) (<-chan expvar, <-chan error) {
+	outChan := make(chan expvar)
+	errChan := make(chan error)
+	ticker := time.NewTicker(period)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				ticker.Stop()
+				close(outChan)
+				return
+			case <-ticker.C:
+				var e expvar
+				ctxWithTimeout, cancel := context.WithTimeout(ctx, period)
+				err := queryExpvar(ctxWithTimeout, &e, serverURL)
+				cancel()
+				if err != nil {
+					errChan <- err
+					return
+				}
+				outChan <- e
+			}
+		}
+	}()
+	return outChan, errChan
+}
```

Comment:
Couple of changes to always stop the ticker and close the channel in the defer call.

Also, the local `expvar` shouldn't be sent over to `outChan` when `err != nil`.

```suggestion
func run(ctx context.Context, serverURL string, period time.Duration) (<-chan expvar, <-chan error) {
	outChan := make(chan expvar)
	errChan := make(chan error)

	go func() {
		ticker := time.NewTicker(period)
		defer func() {
			ticker.Stop()
			close(outChan)
		}()
		query := func(e *expvar) error {
			ctx, cancel := context.WithTimeout(ctx, period)
			defer cancel()
			return queryExpvar(ctx, e, serverURL)
		}
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				var e expvar
				if err := query(&e); err != nil {
					errChan <- err
					continue
				}
				outChan <- e
			}
		}
	}()
	return outChan, errChan
}
```